### DISCUSSION TITLE
Fontfile not working

### DISCUSSION THREAD

**Discussion by Ruchika-K** [libvips/libvips#2994] [Q&A]
Referring to this - https://github.com/libvips/libvips/issues/950
I had tried passing the custom fontfile :

x = pyvips.Image.text("Hello World!",
dpi=300,
font="Lovers Quarrel 32",
fontfile="{AbsolutePath}/LoversQuarrel-Regular.ttf")

This gives image with default font and not as expected.
Is there something I am missing?

**Comment by jcupitt** [comment]
Hi @Ruchika-K,

Maybe the file does not exist on your system? If it's not there, you'll get a warning and it'll use another font instead.

```
$ vips text x.png hello --font banana --fontfile fred.ttf

(vips:1263373): VIPS-WARNING **: 23:18:01.095: unable to load fontfile "fred.ttf"
```

(edit: changed wording to be clearer)

**Reply by Ruchika-K** [reply]
Do we need to install the font on system separately? Is it not platform independent? 
As far as I can see when we add a fontconfig file , libvips loads that file into the fontconfig cache.

**Reply by jcupitt** [reply]
The `fontfile` parameter adds the file directly, so there's no need to install it. For example, I have these font files:

```
$ ls *.ttf
MissFajardose-Regular.ttf    PT_Serif-Web-Bold.ttf    PT_Serif-Web-Regular.ttf
PT_Serif-Web-BoldItalic.ttf  PT_Serif-Web-Italic.ttf
```

But they are not installed, so this:

```
$ vips text x.png hello --font "Miss Fajardose" --dpi 600
```

Makes this:

![x](https://user-images.githubusercontent.com/580843/185256783-38006f9d-9233-41b0-9d47-8921f08ad5f4.png)

It's substituted something like helvetica. But if I enter:

```
$ vips text x.png hello --font "Miss Fajardose" --fontfile MissFajardose-Regular.ttf --dpi 600
```

Now it'll load the crazy font and I get:

![x](https://user-images.githubusercontent.com/580843/185256885-8953761a-b37b-4f53-8623-56aaa220716e.png)

**Reply by Ruchika-K** [reply]
Oh Sorry I misread your statement "file does not exist" as "font does not exist". Hence the above question. Thanks for clarifying that though :)
But regarding the actual issue - The file is present. Also  I do not get such a warning as you have mentioned.

**Reply by Ruchika-K** [reply]
Could this be possible due to any settings in the system? I know that it should be platform independent , but does it use the fontconfig already installed in the system?

**Reply by jcupitt** [reply]
Sorry, I've no idea, you'd need to say what your platform is, what libvips you're using, etc. You've not given enough information for me to be able to reproduce your problem.

To track the issue down, you need to test the system from the bottom layers up. 

1. Can you load the font file in some other program? It might be corrupt.
2. Which libvips do you have? Is is the current stable version? Is it a complete install with a CLI?
3. Are you using conda? If so, I probably can't help, you'd need to ask the people who made the libvips conda binary, since conda will have its own private fontconfig system.
4. Are you using windows? Are there non-ascii characters in the fontfile path?
5. Can you use the libvips CLI to run `vips_text`? Can you use the `--fontfile` argument? Are you using the correct name for the font?
6. Which libvips is pyvips running against? Is it linking to the same one as the CLI interface?
7. Again, do you have the font name correct in python? You may need to match case, spaces, etc.

**Reply by Ruchika-K** [reply]
Hey @jcupitt ,

I tried this as well - 
vips text x.png hello --font "Lovers Quarrel 32" --fontfile LoversQuarrel-Regular.ttf{Have mentioned absolute path}

I am using mac Big Sur Version 11.6.2.
vips version is - vips-8.12.2

Does the fontfile take some relative path ? Since it doesn't throw warning it seems it is loading the fontfile though.

**Reply by jcupitt** [reply]
Ah homebrew. It's probably an issue with the homebrew formula, I'd open a bug there. 

You can set `FC_DEBUG` to get some feedback from fontconfig on font resolution, eg.:

```
FC_DEBUG=8191 vips text ...
```

I think fontconfig is loading the font, but the font rendering system is not using the fontconfig suggestion. Perhaps pangoft on homebrew is using the system font renderer?

**Reply by kleisauke** [reply]
Pango defaults to the CoreText rendering and mapping backend on macOS. You could try to use the Fontconfig backend with the `PANGOCAIRO_BACKEND="fontconfig"` environment variable.

**Reply by Ruchika-K** [reply]
Thanks @jcupitt  

I tried with this debug option. Seems to me fontconfig is picking up the font. Isn't it?:-

Last few snippet from the debug logs-

Final font pattern:
Pattern has 25 elts (size 32)
	family: "Lovers Quarrel"(s)
	familylang: "en"(s)
	style: "Regular"(s)
	stylelang: "en"(s)
	fullname: "Lovers Quarrel"(s)
	fullnamelang: "en"(s)
	slant: 0(i)(s)
	weight: 80(f)(s)
	width: 100(f)(s)
	foundry: "pyrs"(s)
	file: "{absolute-path}/LoversQuarrel-Regular.ttf"(s)
	index: 0(i)(s)
	outline: True(s)
	scalable: True(s)
	charset: 
	0000: 00000000 ffffffff ffffffff 7fffffff 00000000 ffffdfff ffffffff ffffffff
	0001: 00000000 81be0380 03dc001f 01000000 00040000 00000000 00000000 00000000
	0002: 00000000 00000000 00000000 00000000 00000000 00000000 3f0000c0 00000000
	0003: 00000000 00000000 00000000 00000000 00000000 00000000 00000001 00000000
	0020: 77180000 06010047 00000010 00000000 00000000 00001000 00000000 00000000
	0021: 00000000 00000044 00000000 00000000 00000000 00000000 00000000 00000000
	0022: 44068044 00000800 00000100 00000031 00000000 00000000 00000000 00000000
	0025: 00000000 00000000 00000000 00000000 00000000 00000000 00000400 00000000
	00f8: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 80000000
	00fb: 00000006 00000000 00000000 00000000 00000000 00000000 00000000 00000000
(s)
	lang: aa|ay|bi|br|ca|ch|co|da|de|en|es|eu|fj|fo|fr|fur|fy|gd|gl|gv|ho|ia|id|ie|io|is|it|lb|mg|nb|nds|nl|nn|no|nr|oc|om|pt|rm|sma|smj|so|sq|ss|st|sv|sw|tl|ts|uz|vo|wa|xh|yap|zu|an|fil|ht|jv|kj|kwm|li|ms|ng|pap-an|pap-aw|rn|rw|sc|sg|sn|su|za(s)
	fontversion: 65602(i)(s)
	fontformat: "TrueType"(s)
	decorative: False(s)
	postscriptname: "LoversQuarrel-Regular"(s)
	color: False(s)
	symbol: False(s)
	variable: False(s)
	fonthashint: True(s)
	order: 0(i)(s)

**Reply by Ruchika-K** [reply]
Thanks @kleisauke , setting up the env worked!

**Reply by RomanKamlykov** [reply]
> Pango defaults to the CoreText rendering and mapping backend on macOS. You could try to use the Fontconfig backend with the `PANGOCAIRO_BACKEND="fontconfig"` environment variable.

Thanks. I used this env on macOS with the "sharp" package on NodeJS:
```
PANGOCAIRO_BACKEND=fontconfig node main.js
```
```
const sharp = require('sharp');

async function main() {
    await sharp({
        text: {
            text: 'Hello, world!',
            width: 400,
            height: 300,
            font: 'DSEG14 Classic',
            fontfile: '/Users/username/app/DSEG14Classic-Regular.ttf',
        }
    }).toFile('result.png');
}

main();
```
